theano/gof/toolbox.py

```diff
@@ -100,8 +103,15 @@ def __init__(self, old_r, new_r, old_r_val, new_r_val, reason,
         self.reason = reason
         self.old_graph = old_graph
         self.new_graph = new_graph
+        self.full_err = None
+        if isinstance(old_r, str):
+            assert (new_r is None and old_r_val is None and new_r_val is None and
+                    reason is None and old_graph is None and new_graph is None)
+            self.full_err = old_r
```

Comment:
I think it's quite confusing to have essentially two modes for the same Exception.
At least we should have comments explaining why we are doing that here.